File: buckaroo/customizations/histogram.py
import math
import numpy as np


def force_float(n):
    if isinstance(n, np.floating):
        return n.item()
    else:
        return n


def _trim(s: str) -> str:
    """Strip trailing zeros after the decimal point only."""
    if '.' in s:
        s = s.rstrip('0').rstrip('.')
    return s


def fmt_num(value: float, step: float, ref: float) -> str:
    """Format one histogram boundary — SI prefix (K/M/B/T) + step-based precision."""
    if step > 0 and abs(value) < step * 1e-9:
        value = 0.0
    for threshold, suffix in [(1e12, 'T'), (1e9, 'B'), (1e6, 'M'), (1e3, 'K')]:
        if ref >= threshold:
            scaled = value / threshold
            step_s = step / threshold
            dec = max(0, -math.floor(math.log10(step_s)) + 1) if step_s > 0 else 1
            dec = min(dec, 2)
            return _trim(f"{scaled:.{dec}f}") + suffix
    dec = max(0, -math.floor(math.log10(step)) + 1) if step > 0 else 0
    dec = min(dec, 6)
    return _trim(f"{value:.{dec}f}")


def _join_bounds(lo_s: str, hi_s: str) -> str:
    # any negative bound makes '–' ambiguous: the minus sign and en-dash
    # are near-identical glyphs
    sep = '<>' if (lo_s.startswith('-') or hi_s.startswith('-')) else '–'
    return f"{lo_s}{sep}{hi_s}"


def fmt_bucket(lo: float, hi: float, step: float, ref: float) -> str:
    return _join_bounds(fmt_num(lo, step, ref), fmt_num(hi, step, ref))


def fmt_tail_bucket(lo: float, hi: float, step: float) -> str:
    """Format a tail bucket — per-bound SI prefix so an outlier bound
    doesn't drag a small bound to '0K'."""
    return _join_bounds(fmt_num(lo, step, abs(lo)), fmt_num(hi, step, abs(hi)))


def numeric_histogram_labels(endpoints):
    left = endpoints[0]
    labels = []
    min_val = float(endpoints[0])
    max_val = float(endpoints[-1])
    step = (max_val - min_val) / max(len(endpoints) - 1, 1)
    ref = max(abs(min_val), abs(max_val))
    for edge in endpoints[1:]:
        labels.append(fmt_bucket(float(left), float(edge), step, ref))
        left = edge
    return labels

def categorical_dict(len_, val_counts, top_n_positions=7):
    top = min(len(val_counts), top_n_positions)
    top_vals = val_counts.iloc[:top]
        
    rest_vals = val_counts.iloc[top:]
    try:
        histogram = top_vals.to_dict()
    except TypeError:
        top_vals.index = top_vals.index.map(str)
        histogram = top_vals.to_dict()

    full_long_tail = rest_vals.sum()
    unique_count = sum(val_counts == 1)
    long_tail = full_long_tail - unique_count
    if long_tail > 0:
        histogram['longtail'] = np.round((long_tail/len_) * 100,0)
    if unique_count > 0:
        histogram['unique'] = np.round( (unique_count/len_)* 100, 0)
    return histogram    


def categorical_histogram(length:int, val_counts, nan_per:float, top_n_positions=7):
    nan_observation = {'name':'NA', 'NA':np.round(nan_per*100, 0)}
    cd = categorical_dict(length, val_counts, top_n_positions)
    
    histogram = []
    for k,v in cd.items():
        if k in ["longtail", "unique"]:
            continue

        percent = np.round((v/length)*100,0)
        if percent > .3:
            # str(k) important because the key must be a string, not a number or boolean
            """
            Warning: Received `true` for a non-boolean attribute `name`.

            If you want to write it to the DOM, pass a string instead: name="true" or name={value.toString()}.
            path
            Rectangle2@http://localhost:6006/node_modules/...
            """
            histogram.append({'name':str(k), 'cat_pop': percent })

    # I want longtail and unique to come last
    for k,v in cd.items():
        if k in ["longtail", "unique"]:
            obs = {'name': k}
            obs[k] = v
            histogram.append(obs)
    if nan_per > 0.0:
        histogram.append(nan_observation)
    return histogram

# histogram_args = TypedDict('histogram_args', {
#     'meat_histogram': Tuple[npt.NDArray[np.intp], npt.NDArray[Any]],
#     'low_tail': float, 'high_tail':float})

# class Histogram_Args(TypedDict):
#     meat_histogram: Tuple[List[int], List[float]]
#     normalized_populations:List[float]
#     low_tail: float
#     high_tail: float

#def numeric_histogram(histogram_args: Histogram_Args , min_, max_, nan_per):
def numeric_histogram(histogram_args, min_, max_, nan_per):
    low_tail, high_tail = histogram_args['low_tail'], histogram_args['high_tail']
    ret_histo = []
    nan_observation = {'name':'NA', 'NA':np.round(nan_per*100, 0)}
    if nan_per == 1.0:
        return [nan_observation]

    populations, endpoints = histogram_args['meat_histogram']
    labels = numeric_histogram_labels(endpoints)
    normalized_pop = histogram_args['normalized_populations']

    min_f, max_f = force_float(min_), force_float(max_)
    # precision from the meat bucket width — the full min/max range is
    # outlier-inflated and would collapse small tail bounds to '0K'
    e_lo, e_hi = float(endpoints[0]), float(endpoints[-1])
    step = (e_hi - e_lo) / max(len(endpoints) - 1, 1)
    low_label = fmt_tail_bucket(min_f, force_float(low_tail), step)
    ret_histo.append({'name': low_label, 'tail':1})
    for label, pop in zip(labels, normalized_pop):
        ret_histo.append({'name': label, 'population':np.round(pop * 100, 0)})
    high_label = fmt_tail_bucket(force_float(high_tail), max_f, step)
    ret_histo.append({'name': high_label, 'tail':1})
    if nan_per > 0.0:
        ret_histo.append(nan_observation)
    return ret_histo

File: packages/buckaroo-duckdb-node/src/histogram.ts
/**
 * Histogram bars: the `histogram` summary stat the viewer renders through the
 * `displayer: 'histogram'` pinned row.
 *
 * This is a faithful TS port of buckaroo's Python producer
 * (`customizations/histogram.py`) so a DuckDB-backed column renders the same
 * bars as the pandas/polars backends:
 *
 *   - NUMERIC: a low-tail marker, 10 "meat" buckets over the 1st–99th
 *     percentile range (population %), a high-tail marker, then an NA bucket.
 *   - CATEGORICAL: the top-7 categories (`cat_pop` %), then `longtail`,
 *     `unique`, and `NA` buckets.
 *
 * All percentages are on the 0–100 scale and rounded with numpy's
 * round-half-to-even (`np.round`) so the labels match the Python output
 * bit-for-bit. The SQL that produces the raw inputs lives in `histogramSql.ts`.
 */

import type { HistogramBar } from './wireTypes.js';

/** numpy `np.round(x, 0)` — round half to even (banker's rounding). */
export function npRound0(x: number): number {
  const floor = Math.floor(x);
  const diff = x - floor;
  if (diff < 0.5) return floor;
  if (diff > 0.5) return floor + 1;
  // exact .5 → round to the even neighbour
  return floor % 2 === 0 ? floor : floor + 1;
}

/** histogram.py:_trim — strip trailing zeros after the decimal point only. */
function trim(s: string): string {
  if (s.includes('.')) {
    return s.replace(/0+$/, '').replace(/\.$/, '');
  }
  return s;
}

/** histogram.py:fmt_num — SI prefix (K/M/B/T) + step-based precision. */
export function fmtNum(value: number, step: number, ref: number): string {
  if (step > 0 && Math.abs(value) < step * 1e-9) {
    value = 0.0;
  }
  const tiers: Array<[number, string]> = [
    [1e12, 'T'],
    [1e9, 'B'],
    [1e6, 'M'],
    [1e3, 'K'],
  ];
  for (const [threshold, suffix] of tiers) {
    if (ref >= threshold) {
      const scaled = value / threshold;
      const stepS = step / threshold;
      let dec = stepS > 0 ? Math.max(0, -Math.floor(Math.log10(stepS)) + 1) : 1;
      dec = Math.min(dec, 2);
      return trim(scaled.toFixed(dec)) + suffix;
    }
  }
  let dec = step > 0 ? Math.max(0, -Math.floor(Math.log10(step)) + 1) : 0;
  dec = Math.min(dec, 6);
  return trim(value.toFixed(dec));
}

/**
 * histogram.py:_join_bounds — any negative bound switches the separator to
 * `<>`; the minus sign and en-dash are near-identical glyphs and `-0.5–0.5`
 * reads as a double dash.
 */
function joinBounds(loS: string, hiS: string): string {
  const sep = loS.startsWith('-') || hiS.startsWith('-') ? '<>' : '–';
  return `${loS}${sep}${hiS}`;
}

/** histogram.py:fmt_bucket */
export function fmtBucket(lo: number, hi: number, step: number, ref: number): string {
  return joinBounds(fmtNum(lo, step, ref), fmtNum(hi, step, ref));
}

/**
 * histogram.py:fmt_tail_bucket — per-bound SI prefix (ref = |bound|) so a far
 * outlier bound doesn't drag a small bound down to '0K'.
 */
function fmtTailBucket(lo: number, hi: number, step: number): string {
  return joinBounds(fmtNum(lo, step, Math.abs(lo)), fmtNum(hi, step, Math.abs(hi)));
}

/** histogram.py:numeric_histogram_labels — one label per bucket from the edges. */
export function numericHistogramLabels(endpoints: number[]): string[] {
  let left = endpoints[0];
  const labels: string[] = [];
  const minVal = endpoints[0];
  const maxVal = endpoints[endpoints.length - 1];
  const step = (maxVal - minVal) / Math.max(endpoints.length - 1, 1);
  const ref = Math.max(Math.abs(minVal), Math.abs(maxVal));
  for (const edge of endpoints.slice(1)) {
    labels.push(fmtBucket(left, edge, step, ref));
    left = edge;
  }
  return labels;
}

/**
 * The raw numeric-histogram inputs, mirroring Python's `histogram_args`:
 *   - `meatCounts`        — np.histogram bin counts (10 bins).
 *   - `endpoints`         — the 11 bin edges (linspace over the meat range).
 *   - `lowTail`/`highTail`— the 1st/99th percentile cut points.
 * `normalizedPopulations` is derived here (counts / sum) so callers only carry
 * the SQL-shaped values.
 */
export interface NumericHistogramArgs {
  meatCounts: number[];
  endpoints: number[];
  lowTail: number;
  highTail: number;
}

/** histogram.py:numeric_histogram */
export function numericHistogram(
  args: NumericHistogramArgs,
  min: number,
  max: number,
  nanPer: number,
): HistogramBar[] {
  const naObs: HistogramBar = { name: 'NA', NA: npRound0(nanPer * 100) };
  if (nanPer === 1.0) return [naObs];

  const { meatCounts, endpoints, lowTail, highTail } = args;
  const total = meatCounts.reduce((a, b) => a + b, 0);
  const normalizedPop = meatCounts.map((c) => (total > 0 ? c / total : 0));
  const labels = numericHistogramLabels(endpoints);

  const eLo = endpoints[0];
  const eHi = endpoints[endpoints.length - 1];
  const step = (eHi - eLo) / Math.max(endpoints.length - 1, 1);

  const histo: HistogramBar[] = [];
  histo.push({ name: fmtTailBucket(min, lowTail, step), tail: 1 });
  labels.forEach((label, i) => {
    histo.push({ name: label, population: npRound0(normalizedPop[i] * 100) });
  });
  histo.push({ name: fmtTailBucket(highTail, max, step), tail: 1 });
  if (nanPer > 0.0) histo.push(naObs);
  return histo;
}

/** A distinct value and its row count, as returned by the value-counts query. */
export interface ValueCount {
  name: string;
  count: number;
}

/**
 * histogram.py:categorical_histogram (with categorical_dict inlined).
 *
 * `top` is the top-N value counts (already sorted desc); `restSum` is the
 * summed count of every other distinct value; `uniqueCount` is the number of
 * distinct values that occur exactly once (over the whole column). `length`
 * includes nulls.
 */
export function categoricalHistogram(
  length: number,
  top: ValueCount[],
  restSum: number,
  uniqueCount: number,
  nanPer: number,
): HistogramBar[] {
  const histo: HistogramBar[] = [];
  for (const { name, count } of top) {
    const percent = npRound0((count / length) * 100);
    if (percent > 0.3) {
      histo.push({ name: String(name), cat_pop: percent });
    }
  }
  const longTail = restSum - uniqueCount;
  if (longTail > 0) {
    histo.push({ name: 'longtail', longtail: npRound0((longTail / length) * 100) });
  }
  if (uniqueCount > 0) {
    histo.push({ name: 'unique', unique: npRound0((uniqueCount / length) * 100) });
  }
  if (nanPer > 0.0) {
    histo.push({ name: 'NA', NA: npRound0(nanPer * 100) });
  }
  return histo;
}

/** The categorical inputs `buildHistogram` needs (see `parseCategorical`). */
export interface CategoricalArgs {
  top: ValueCount[];
  restSum: number;
  uniqueCount: number;
}

/**
 * histogram.py:histogram — pick the numeric path only when the column is
 * numeric, has more than 5 distinct values, valid histogram args, and the
 * resulting numeric histogram has more than 5 bars; otherwise fall back to the
 * categorical histogram.
 *
 * The numeric and categorical inputs are fetched lazily so the categorical
 * query only runs when the numeric path doesn't win — matching histogram.py's
 * short-circuit. This is the single dispatcher: the production pipeline
 * (histogramSql.ts:computeColumnHistogram) supplies SQL-backed fetchers, the
 * unit tests supply in-memory ones, so the same branch logic is what ships.
 */
export async function buildHistogram(opts: {
  isNumeric: boolean;
  distinctCount: number;
  length: number;
  nanPer: number;
  min: number | null;
  max: number | null;
  fetchNumericArgs: () => Promise<NumericHistogramArgs | null>;
  fetchCategorical: () => Promise<CategoricalArgs>;
}): Promise<HistogramBar[]> {
  const { isNumeric, distinctCount, length, nanPer, min, max, fetchNumericArgs, fetchCategorical } =
    opts;
  if (isNumeric && distinctCount > 5 && min !== null && max !== null) {
    const numericArgs = await fetchNumericArgs();
    if (numericArgs) {
      const temp = numericHistogram(numericArgs, min, max, nanPer);
      if (temp.length > 5) return temp;
    }
  }
  const categorical = await fetchCategorical();
  return categoricalHistogram(length, categorical.top, categorical.restSum, categorical.uniqueCount, nanPer);
}
